cibuildwheel/platforms/android.py

```diff
@@ -0,0 +1,625 @@
+
+
+def setup_env(
+    config: PythonConfiguration, build_options: BuildOptions, build_path: Path, python_dir: Path
+) -> tuple[dict[str, str], dict[str, str]]:
+    """
+    Returns two environment dicts, both pointing at the same virtual environment:
+
+    * build_env, which uses the environment normally.
+    * android_env, which uses the environment while simulating running on Android.
+    """
+    log.step("Setting up build environment...")
+
+    # Create virtual environment
+    python_exe = create_python_build_standalone_environment(
+        config.version, build_path, CIBW_CACHE_PATH
+    )
+    venv_dir = build_path / "venv"
+    dependency_constraint = build_options.dependency_constraints.get_for_python_version(
+        version=config.version, tmp_dir=build_path
+    )
+    build_env = virtualenv(
+        config.version, python_exe, venv_dir, dependency_constraint, use_uv=False
+    )
+    create_cmake_toolchain(config, build_path, python_dir, build_env)
+
+    # Apply custom environment variables, and check environment is still valid
+    build_env = build_options.environment.as_dictionary(build_env)
+    build_env["PIP_DISABLE_PIP_VERSION_CHECK"] = "1"
+    for command in ["python", "pip"]:
+        which = call("which", command, env=build_env, capture_stdout=True).strip()
+        if which != f"{venv_dir}/bin/{command}":
+            msg = (
+                f"{command} available on PATH doesn't match our installed instance. If you "
+                f"have modified PATH, ensure that you don't overwrite cibuildwheel's entry "
+                f"or insert {command} above it."
+            )
+            raise errors.FatalError(msg)
+        call(command, "--version", env=build_env)
+
+    # Construct an altered environment which simulates running on Android.
+    android_env = setup_android_env(python_dir, venv_dir, build_env)
+
+    # Install build tools
+    build_frontend = build_options.build_frontend
+    if build_frontend.name != "build":
+        msg = "Android requires the build frontend to be 'build'"
+        raise errors.FatalError(msg)
+    call("pip", "install", "build", *constraint_flags(dependency_constraint), env=build_env)
+
+    # Install build-time requirements. These must be installed for the build platform,
+    # but queried while simulating Android. The `build` CLI doesn't support this
+    # combination, so we use its API to query the requirements, but install them
+    # ourselves. We'll later run `build` in the same environment, passing the
+    # `--no-isolation` option.
+    class AndroidEnv(IsolatedEnv):
+        @property
+        def python_executable(self) -> str:
+            return f"{venv_dir}/bin/python"
+
+        def make_extra_environ(self) -> dict[str, str]:
+            return android_env
+
+    pb = ProjectBuilder.from_isolated_env(AndroidEnv(), build_options.package_dir)
+    if pb.build_system_requires:
+        call("pip", "install", *pb.build_system_requires, env=build_env)
```

Comment:
So I understand this, if a package had an android wheel, `ProjectBuilder` would find that file and return a full URL to it here?

But the `pip install` command is executed in the build_env. So if it was compiling from a `tar.gz` it would build for the wrong platform in that case?

Reply:
`ProjectBuilder` only returns requirement specifiers; it's `pip install` that resolves them into URLs. I've updated the comment to clarify the distinction between the Android platform and build platform in this function.